### DISCUSSION TITLE
Deploying Code Analysis package as Nuget that depends on another Nuget

### DISCUSSION THREAD

**Discussion by myblindy** [dotnet/roslyn#39682] [General]
**Version Used**: 16.3.5

I have a code analysis package that works when run through the VSIX debugger, but when I try to publish the nuget and reference it, i get the following error:

    Analyzer 'RIDAnalyzer.RIDAnalyzerAnalyzer' threw an exception of type 'System.IO.FileNotFoundException' with message 'Could not load file or assembly 'MoreLinq, Version=3.2.0.0, Culture=neutral, PublicKeyToken=384d532d7e88985d' or one of its dependencies. The system cannot find the file specified.'.	

My .csproj file looks like this:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <IncludeBuildOutput>false</IncludeBuildOutput>
    <GeneratePackageOnBuild>True</GeneratePackageOnBuild>
  </PropertyGroup>
  <ItemGroup>
    <Compile Remove="bin\**" />
    <EmbeddedResource Remove="bin\**" />
    <None Remove="bin\**" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="2.9.7" PrivateAssets="all" />
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="3.3.1" PrivateAssets="all" />
    <PackageReference Include="morelinq" Version="3.2.0" />
    <PackageReference Update="NETStandard.Library" PrivateAssets="all" />
  </ItemGroup>

  <ItemGroup>
    <Compile Update="Resources.Designer.cs" DesignTime="True" AutoGen="True" DependentUpon="Resources.resx" />
    <EmbeddedResource Update="Resources.resx" Generator="ResXFileCodeGenerator" LastGenOutput="Resources.Designer.cs" />
  </ItemGroup>

  <ItemGroup>
    <None Update="tools\*.ps1" CopyToOutputDirectory="Always" Pack="true" PackagePath="" />
    <None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
  </ItemGroup>

</Project>
```

And it only generates the executable and a reference on `morelinq`, the nuget dependency. I see the dependency in the Nuget package manager, and it installs fine (supposedly) -- including showing up in the "analyzers" tab in my project, but when I try to use the analyzer I get the warning above and nothing else happens.

**Comment by mavasani** [comment]
It seems that your analyzer assembly depends on morelinq.dll. You need to package the analyzer dependency into your package and it should be marked as an analyzer assembly itself. This is needed to ensure that when analyzers are executed with shadow copying, the required dependency can be found at runtime.

**Comment by myblindy** [comment]
Okay, but how do I do that? What I ended up doing was copy the dll manually in a `deps` folder and hard reference it in my project, then add the link to the `.csproj` file as a pack reference.

This can't _possibly_ be the only way to tell VS to just get a nuget package and use it, right?

**Comment by rolan18** [comment]
I was able to get it working by adding the following to the project file:
```
  <ItemGroup>
    <PackageReference Include="Package.Name" Version="1.0.0" PrivateAssets="all" GeneratePathProperty="true"/>
<None Include="$(PkgPackage_Name)\lib\netstandard2.0\*.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
</ItemGroup>
```
I'm not going to claim that I know if every property is necessary, but this will put the dlls from the package into your analyzers folder in the nupkg.
